components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/autofill/LoginStorageDelegate.kt

```diff
@@ -0,0 +1,121 @@
+}
+
+/**
+ * [LoginDelegate /* TODO UPDATE */] implementation.
+ *
+ * When GV needs to update or retrieve information about stored logins, it will call into this
+ * class.
+ *
+ * App will have to instantiate this and set it on the runtime and pass in the [LoginsStorage] and
+ * [SecureAbove22Preferences] with a key that conforms to "passwords".  This lets GV delegate to
+ * the client for storage.
+ */
+class LoginStorageDelegate(
+    private val loginStorage: LoginsStorage,
+    private val keyStore: SecureAbove22Preferences
+) : LoginDelegate {
+    override fun onLoginUsed(login: Login) {
+        val passwordsKey = keyStore.getString(PASSWORDS_KEY)
+        val guid = login.guid
+        // TODO should this short on an empty string?
+        if (passwordsKey == null || guid == null || guid.isEmpty()) return
+        with(loginStorage) {
+            ensureUnlocked(passwordsKey)
+            touch(guid)
+            lock()
+        }
+    }
+
+    override fun onFetchLogins(domain: String): GeckoResult<Array<Login>> {
+        val passwordsKey =
+            keyStore.getString(PASSWORDS_KEY) ?: return GeckoResult.fromValue(arrayOf())
+        loginStorage.ensureUnlocked(passwordsKey).also {
+            val result = GeckoResult.fromValue(loginStorage.getByHostname(domain).map { // TODO getByHostname -> getByBaseDomain
+                Login(
+                    guid = it.id,
+                    origin = it.hostname,
+                    formActionOrigin = it.formSubmitURL,
+                    httpRealm = it.httpRealm,
+                    username = it.username ?: "", // TODO this should be nonnull after an incoming AS update
+                    password = it.password
+                )
+            }.toTypedArray())
+            loginStorage.lock()
+            return result
+        }
+    }
+
+    // TODO double check that we're locking correctly.  Also, does this need to be synchronized?
+    // Request to save or update the given login.
+    @Synchronized
+    override fun onLoginSave(login: Login) {
+        val passwordsKey = keyStore.getString(PASSWORDS_KEY) ?: return // TODO should we fail fast if this is null?
+        loginStorage.ensureUnlocked(passwordsKey)
+        val guid = login.guid
+        val serverPassword = guid?.let { loginStorage.get(it) }
+
+        if (guid != null && serverPassword != null) {
+            loginStorage.update(serverPassword.mergeWithLogin(login))
+        } else {
+            loginStorage.add(
+                ServerPassword(
+                    id = "", // TODO ask if this is right? Pass empty string if we don't have it?  If so, ask for it to be null
+                    username = login.username,
+                    password = login.password,
+                    hostname = login.origin,
+                    formSubmitURL = login.formActionOrigin,
+                    httpRealm = login.httpRealm,
+                    usernameField = "", // TODO This seems problematic. Run it by Vlad
```

Comment:
@vladikoff these types are showing up here. If we're adding a new ServerPassword I'm assuming this is fine? We're not getting this info passed from GV AFAIK

Reply:
This is gonna be weird now, because we will end up with logins without a field. This might mean Firefox Desktop won't be able to fill these?

@thomcc thoughts?

Reply:
@mnoorenberghe do you know where we should be getting `usernameField`?  I imagine that should come from GV, but I don't see it [in the proposed Login class](https://docs.google.com/document/d/1VrYeAASErlT7FwagxPmMzDnRewCUVnp-CVND8zOO5hw/edit#heading=h.pjqhqgqq5iv6).

Reply:
No platforms actually use `usernameField`… it does get populated in toolkit code but never used later. We haven't made the judgement call to remove it or start using it. Setting it to an empty string for now is fine with me, same for `passwordField`